File: addon/behavior_pack/scripts/actions/pet.js
export const defaultParticle = "beatpet:petpet";
export const defaultAnimation = "animation.beatpet.petpet";

/**
 * @param {import('@minecraft/server').Entity} entity
 * @param {string} particle
 * @param {string} animation
 */
export function pet(entity, particle = defaultParticle, animation = defaultAnimation) {
    const location = entity.getHeadLocation();
    location.y += .3;
    entity.dimension.spawnParticle(particle, location);
    entity.playAnimation(animation, {
        blendOutTime: 0,
        controller: 'anim1',
        nextState: 'none',
        stopExpression: '',
    });

    if (entity.typeId === 'minecraft:player')
        entity.runCommand('camerashake add @s 1 0.2 positional');
}

File: allay/src/main/kotlin/io/github/minebuilders/beatpet/actions/Pet.kt
package io.github.minebuilders.beatpet.actions

import org.allaymc.api.entity.Entity
import org.allaymc.api.entity.data.EntityAnimation
import org.allaymc.api.world.particle.CustomParticle
import org.joml.Vector3d

const val DEFAULT_PARTICLE = "beatpet:petpet"
const val DEFAULT_ANIMATION = "animation.beatpet.petpet"

fun Entity.pet(
    template: String = DEFAULT_PARTICLE,
    animation: String = DEFAULT_ANIMATION,
) {
    val location = Vector3d(location)
    location.y += (offsetAABBForCollisionCheck.maxY() - offsetAABBForCollisionCheck.minY())
    location.y += 0.3f
    dimension.addParticle(location, CustomParticle(template))

    applyAnimation(EntityAnimation(animation, "none", "anim1", ""))

    // TODO: camera shake if this is player
}
